### DISCUSSION TITLE
Automated Backups of Nightscout in Google Cloud

### DISCUSSION THREAD

**Discussion by motinis** [NightscoutFoundation/xDrip#2877] [Q&A] [answered]
Hi All,

Is there anyway to automate the backup process? For example, I wouldn't mind use the scheduled snapshots but I guess this would be better to do with [guest flush](https://cloud.google.com/compute/docs/disks/creating-linux-application-consistent-pd-snapshots) to ensure the data is consistent.

Has anybody done this? Any pointers?

Thanks!

**Comment by Navid200** [comment]
I don't know how to automatically upload a backup from the virtual machine to Google Drive.
If I figure that out, I can set it up.

Without that, any backup will still be on the virtual machine itself.  
The way I see this, the whole idea of the backup is to protect the user against an accidental deletion of the virtual machine.   If the backup is on the virtual machine, in case of an accidental deletion, the backups will be deleted too.   
So, as long as we don't have a way to transfer data, for free, from a virtual machine to Google Drive, we have no solution.

**Comment by motinis** [comment]
So - there's also [5GB free worth of storage](https://cloud.google.com/storage/pricing#cloud-storage-always-free) in Google Cloud Storage. In any case, the costs there are fairly minimal. I think it would probably be simplest to have this be driven from the VM side - in other words have a cron job periodically upload a file into a GCS bucket. I think that you would generally just use versioning and "overwrite" the previous file (you can configure how many versions it keeps and after how long it will delete old versions). I think this would be the simplest way.

**Reply by Navid200** [reply]
This is the only thing I have always considered to be a possible problem, the possibility of the machine being accidentally deleted.  
So, I am very interested to pursue this.  Thanks for the idea.

**Reply by motinis** [reply]
Updated my comment above with the link detailing the Free limits (e.g., I would think one would want this to be in a different region than the VM for example just to better handle DR situations)

**Reply by motinis** [reply]
@Navid200 - should we create a separate issue to track this?

**Reply by Navid200** [reply]
Google Cloud Nightscout is supported by the xDrip team.  But, it's not really in this repository.
I have asked the developers and am waiting for their response.
I wonder if we should move the discussions as well as issues to their own repository.
I hate to have an issue in this repository that is not directly related to xDrip.

Thanks for bringing this up.  Please be patient.

**Comment by motinis** [comment]
You mean this repo (and branch specifically) here: https://github.com/jamorham/nightscout-vps/tree/vps-1 ?

I think it would make sense for sure to open the issue there considering that's where all the code is for this

**Reply by motinis** [reply]
The amount of work here doesn't seem too big - but I totally appreciate everyone is volunteering their time (and it is definitely at least a few hours of set up, testing and documenting)

Documentation in terms of creating keys:
https://developer.bitmovin.com/encoding/docs/how-can-i-create-accesssecret-keys-for-google-cloud-storage

[s3cmd](https://tecadmin.net/install-s3cmd-manage-amazon-s3-buckets/) would need to be installed and configured

And then scripting in terms of scheduling a backup and copying of the file to the bucket

**Reply by motinis** [reply, marked as answer]
One can also do it without any s3 setup; the following works for me (saved it as ~/s3-backup.sh):

```
#! /bin/bash
source ~/setup-s3.sh
s3cmd --disable-multipart --host=storage.googleapis.com --host-bucket="%(bucket).storage.googleapis.com" --access_key "${S3_ACCESS_KEY}"  --secret_key "${S3_SECRET_KEY}" put ~/gcp_nightscout_backup s3://${S3_BUCKET}/gcp_nightscout_backup
```

where I set the environment variables inside the ~/setup-s3.sh script

In my case under interoperability I setup the project as the default (not sure if this is necessary or not), and am using an HMAC key under a service account(the same one as for compute). One doesn't need to have it under the user account

**Reply by Navid200** [reply]
Yes, that's the repository and branch.

Automated backup of the database would be a significant improvement to this project in my opinion.

How much space does it occupy?
How does it line up with free tier?  In other words, what is the cost?
How easy is it to include it in the installation script so that it is installed for everyone during initial install?
How easy would it be to add it as an incremental update for those who already have an existing setup?

**Reply by motinis** [reply]
So space depends on the size of the backup - I think it grows by about 5MB / month of activity. So one can store roughly 1000 months of data in the free tier. Of course, maintaining multiple versions reduces that; e.g. if one wants to keep the last 5 versions then I think it would drop to 200 months / version (I don't think it stores things as deltas). I'm not sure how one would go about "trimming" old data from the mongodb, but that should perhaps be an option. Pricing though is cheap anyway; generally speaking 2 cents / GB / month (see https://cloud.google.com/storage/pricing#price-tables and the free tier regions are US-WEST1, US-CENTRAL1, and US-EAST1), so even if one doesn't trim away old data, it's going to be fairly cheap.

I don't think it would be too hard to have another menu item for automated backup, under which you could configure things (which bucket, the access keys, how often to backup). For incremental update that script could install s3cmd if its not already installed (very simple - `sudo apt-get install s3cmd`). We could consider it creating a bucket automatically for you (it's not hard) but people should definitely review where it's free - i would also generally recommend using a different region than the location of the VM since that'll be safer (although that might (?) cut into network egress, so it should be definitely up to people to decide on their own).

**Reply by Navid200** [reply]
Why did we not worry about backing up the database when we used Heroku?

**Reply by Navid200** [reply]
Isn't what we need just to backup once a month to our computer?  We already have a backup utility that backs up the database and variables into  a single file.  It needs to be downloaded to the computer though.

**Reply by Navid200** [reply]
Now, if someone wants to put that backup file in their Google Drive instead of on their local computer disk, they can.

**Reply by motinis** [reply]
I literally take the file produced by the backup option and have a copy on my computer and then upload to the google cloud storage. Having now a script there just makes it easier; the next step is just adding some cron job around it.

I think it's a user decision for how often they want to backup their data and the retention they want. How much data you are willing to lose in the case something happens to the VM (and how much you are willing to pay for that) is the question.

I can't speak about how the data was being backed up on Heroku.

**Comment by motinis** [comment]
I also had to install cron and then setup a crontab to do daily backups to my GCS. This has been working great. Currently I'm seeing the data grow by about .1 - .2 MB per day, so consistent with 5MB per month. At some point I will need to trim out old data from the DB (or start paying, albeit the prices are cheap) but that isn't any different than on any other provider.
